### PR TITLE
#150 jsonForm database query

### LINKED ISSUE #150: Json Forms (3) -- query document endpoint
Part of Epic #81

### PR DESCRIPTION
**Pull request by CarlosNZ**
Fix #150 

Basic implementation of database query in useJsonForms.

I'm not super happy with the url, or the resulting breadcrumb -- suggestions welcome.

To test: this database has working jsonSchema, uiSchema and Patient data (for 2 patients): `PR227_PatientDocumentQuery.zip` in Google Drive share folder

Click on patient, see the Detail page for that patient load with their document data

### CODE REVIEW

File: client/packages/common/src/hooks/useConfirmOnLeaving/useConfirmOnLeaving.ts [hunk 1]
```diff
@@ -4,7 +4,7 @@ import { useEffect } from 'react';
 // then this hook won't catch the navigation action
 // you will need to use the `Prompt` component instead ( or usePrompt or useBlocker )
 // however, these weren't implemented in react-router-dom v6 at the time of implementation
-export const useConfirmOnLeaving = (isUnsaved: boolean) => {
+export const useConfirmOnLeaving = (isUnsaved: boolean | undefined) => {
```
Comment (CarlosNZ) [inline client/packages/common/src/hooks/useConfirmOnLeaving/useConfirmOnLeaving.ts:7]: Just to allow `isDirty` to be undefined initially

File: client/packages/common/src/ui/forms/JsonForms/useJsonForms.tsx [hunk 2]
```diff
@@ -80,17 +89,37 @@ interface JsonFormOptions {
...
   const navigate = useNavigate();
-  const { isDirty, setIsDirty } = useDirtyCheck();
+
+  useEffect(() => {
+    if (!databaseResponse) return;
+
+    const {
+      data,
+      // @ts-ignore
+      documentRegistry: { jsonSchema, uiSchema },
```
Comment (CarlosNZ) [inline client/packages/common/src/ui/forms/JsonForms/useJsonForms.tsx:113]: Not intending to leave this here, but needed it for now to compile. Typescript is complaining that:
```
Property 'jsonSchema' does not exist on type '{ __typename: "DocumentRegistryNode"; uiSchemaType: string; documentType: string; context: DocumentRegistryNodeContext; jsonSchema: any; uiSchema: any; } | null | undefined'.ts(2339)
```
But those fields are clearly defined in the auto-generated types. Any ideas?

File: server/server/src/sync/init_programs_data.rs [hunk 3]
```diff
@@ -178,7 +181,7 @@ pub fn init_program_data(
         id: patient_schema_id.clone(),
         r#type: "JsonForms".to_string(),
         json_schema: serde_json::from_str(PATIENT_SCHEMA).unwrap(),
-        ui_schema: json!({}),
+        ui_schema: serde_json::from_str(PATIENT_UI_SCHEMA).unwrap(),
```
Comment (CarlosNZ) [inline server/server/src/sync/init_programs_data.rs:184]: Is this what you meant, @clemens-msupply ?

File: server/service/src/document/schemas/patient.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
-  "$ref": "#/definitions/Patient",
```
Comment (CarlosNZ) [inline server/service/src/document/schemas/patient.json:2]: This line causes infinite recursion in JSONForms, so needed to remove.

File: server/server/src/sync/init_programs_data.rs (hunk 3 above)
Comment (clemens-msupply) [reply server/server/src/sync/init_programs_data.rs:184]: yes!

File: client/packages/common/src/ui/forms/JsonForms/useJsonForms.tsx (hunk 2 above)
Comment (clemens-msupply) [reply client/packages/common/src/ui/forms/JsonForms/useJsonForms.tsx:113]: don't know, maybe the linter doesn't like `any` types...

**Review by clemens-msupply — COMMENTED** [review]
I need the patient.json json schema in service because I need the data type there. However, patient_uiSchema.json is only needed for development. For this reason I would put it next to the init_programs_data.rs file.

File: server/service/src/document/schemas/patient.json [hunk 5]
```diff
@@ -174,14 +170,26 @@
...
+              "field1": {
+                "description": "First item",
+                "type": "string"
+              },
+              "field2": {
+                "description": "Second item",
+                "type": "string"
+              }
+            }
+          }
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply server/service/src/document/schemas/patient.json:189]: Why does Rust even need to know what the structure of a doc schema is -- it could be anything and change at any time right?

It doesn't need to be part of any "real" patient data or anything, it's just there so I can test the range of possibilities in the JSON Forms component. In this case, it refers to a field that is missing in the example patient data documents, so we can ensure that the form builder displays the fields correctly anyway (not obvious, since the Array component iterates over existing data element, so this just ensures we've remembered to create an empty value for this field in the data object if its not defined)

[the remark this answers is not in the document]
Comment (CarlosNZ) [reply server/service/src/document/schemas/patient.json:189]: Ah, I see from your later comment that you're using this schema to create the Patient data type. I don't know that we should rely on a document schema for this. A real-word patient document might contain a whole bunch of fields that are not part of the "core" patient data type, wouldn't it?

[the remark this answers is not in the document]
Comment (clemens-msupply) [reply server/service/src/document/schemas/patient.json:189]: There is a base patient schema that all derived schema should be compatible to. I use this base patient schema to derive the rust type, e.g. to parse the json data as a patient object and then fill the names table from it.

When updating a patient I would actually have to check that the data fulfills the base patient schema and the actual patient schema for the data.

So you could actually put your custom patient schema into the schema table during initialization and that should work as well (don't touch the schema in service).

[the remark this answers is not in the document]
Comment (CarlosNZ) [reply server/service/src/document/schemas/patient.json:189]: > So you could actually put your custom patient schema into the schema table during initialization and that should work as well (don't touch the schema in service).

Yeah, that's probably better for now. I'll change it.

We'll sort out a "definitive" json schema in the next few days -- the current one is intended to test a few edge cases.

**Comment by clemens-msupply** [comment]
@CarlosNZ I committed a fix for the default patient schema and I now initialize the data using your custom patient schema.

I also added a comment about the schema validation.

Moreover I sneaked in a fix for showing disabled UI elements as disabled (used it for the Patient id).

**Review by clemens-msupply — APPROVED** [review]

File: client/packages/common/src/hooks/useConfirmOnLeaving/useConfirmOnLeaving.ts (hunk 1 above)
Comment (mark-prins) [inline client/packages/common/src/hooks/useConfirmOnLeaving/useConfirmOnLeaving.ts:7]: :nit
```suggestion
export const useConfirmOnLeaving = (isUnsaved?: boolean) => {
```

File: client/packages/common/src/ui/forms/JsonForms/api/api.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  get: {
+    byDocName: async (name: string): Promise<DocumentFragment> => {
+      const result = await sdk.documentByName({ name, storeId });
+      const document = result?.document;
+
+      if (document?.__typename === 'DocumentNode') {
+        return document;
+      } else {
+        throw new Error('Error querying document');
+      }
```
Comment (mark-prins) [inline client/packages/common/src/ui/forms/JsonForms/api/api.ts:13]: :nit the else isn't required and differs from the pattern elsewhere

```suggestion
      if (document?.__typename === 'DocumentNode') {
        return document;
      }

      throw new Error('Error querying document');      
```

File: client/packages/common/src/ui/forms/JsonForms/api/hooks/document/useDocument.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import { useQuery } from '@openmsupply-client/common';
+import { useDocumentApi } from '../utils/useDocumentApi';
+
+export const useDocument = (docName: string) => {
```
Comment (mark-prins) [inline client/packages/common/src/ui/forms/JsonForms/api/hooks/document/useDocument.ts:4]: :nit I would use `name` rather than `docName`. This is in the `useDocument` hook, so the document part is expected already 🤷  The `doc` doesn't add much; if needed, I'd use `documentName`

File: client/packages/common/src/ui/forms/JsonForms/api/hooks/utils/useDocumentApi.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import { useGql, useAuthContext } from '@openmsupply-client/common';
+import { getDocumentQueries } from '../../api';
+import { getSdk } from '../../operations.generated';
+
+export const useDocumentApi = () => {
+  const { storeId } = useAuthContext();
+  const keys = {
+    detail: (id: string) => [storeId, id] as const,
```
Comment (mark-prins) [inline client/packages/common/src/ui/forms/JsonForms/api/hooks/utils/useDocumentApi.ts:8]: :nit the usage is fetching by name, so this should be `name: string` rather than `id`

strictly speaking, should be `docName` to match usage, but, as noted, I'd prefer to simplify `docName` anyway

File: client/packages/common/src/ui/forms/JsonForms/useJsonForms.tsx [hunk 9]
```diff
@@ -26,38 +31,42 @@ import {
...
+      schema={jsonSchema}
       uischema={uiSchema}
       data={data}
       renderers={renderers}
       // cells={materialCells}
       onChange={({ errors, data }) => {
         setData(data);
-        if (errors && errors.length) console.warn('Errors: ', errors);
+        if (errors && errors.length) {
+          setError(String(errors));
```
Comment (mark-prins) [inline client/packages/common/src/ui/forms/JsonForms/useJsonForms.tsx:67]: errors is an array (`ErrorObject[]`) - is `String(errors)` the best option here?
perhaps an `errors?.map(({message}) => message ?? '').join(', ')` or errors[0] or an error array formatter? 
or `JSON.stringify(errors)`?

File: client/packages/common/src/ui/forms/JsonForms/useJsonForms.tsx [hunk 10]
```diff
@@ -26,38 +31,42 @@ import {
...
 }: JsonFormsComponentProps) => {
   return (
     <JsonForms
-      schema={schema}
+      schema={jsonSchema}
       uischema={uiSchema}
       data={data}
       renderers={renderers}
       // cells={materialCells}
       onChange={({ errors, data }) => {
```
Comment (mark-prins) [inline client/packages/common/src/ui/forms/JsonForms/useJsonForms.tsx:64]: ts is complaining about this for me. not for you?

File: client/packages/common/src/ui/forms/JsonForms/useJsonForms.tsx [hunk 11]
```diff
@@ -80,17 +89,37 @@ interface JsonFormOptions {
...
-  const { isDirty, setIsDirty } = useDirtyCheck();
+
+  useEffect(() => {
+    if (!databaseResponse) return;
+
+    const {
+      data,
+      // @ts-ignore
+      documentRegistry: { jsonSchema, uiSchema },
+    } = databaseResponse;
```
Comment (mark-prins) [inline client/packages/common/src/ui/forms/JsonForms/useJsonForms.tsx:114]: It's complaining because `documentRegistry` can be null | undefined, in which case the destructuring will fail.

```suggestion
    if (!databaseResponse) return;

    const { data, documentRegistry } = databaseResponse;
    const { jsonSchema, uiSchema } = documentRegistry ?? {};

```

File: client/packages/common/src/ui/forms/JsonForms/useJsonForms.tsx [hunk 12]
```diff
@@ -179,17 +213,23 @@ export const useJsonForms = (
...
         <ScrollFix />
-        <FormComponent
-          data={data}
-          setData={updateData}
-          // setError={setError}
-          renderers={renderers}
-        />
-        {showButtonPanel && <ButtonPanel />}
+        {isLoading || !data ? (
+          <Pulse />
```
Comment (mark-prins) [inline client/packages/common/src/ui/forms/JsonForms/useJsonForms.tsx:217]: I prefer the pulse.. but would still use the `<BasicSpinner />` because has the loading text

File: client/packages/system/src/Patient/ListView/ListView.tsx [hunk 13]
```diff
@@ -70,7 +70,7 @@ const PatientListComponent: FC = () => {
         isError={isError}
         onRowClick={row => {
           if (!row.id || !row.document?.name || !row.document?.type) alert();
-          else navigate(`/patients/${row.id}`);
+          else navigate(`/patients/detail?doc=${row.document.name}`);
```
Comment (mark-prins) [inline client/packages/system/src/Patient/ListView/ListView.tsx:73]: ok, so it's nit-picky, but the url makes me sad.

I don't like having the implementation specific detail of `doc=` in there 🤷 but I understand that there's a need for it to be in there

As far as the order & contents: I would  prefer `/patients/${row.id}?doc=${document.type}`
We have 
- `patients` explaining the section
- [patient id] to get the detail
- some specifics in the query

this is the format elsewhere, for example `/distribution/outbound-shipment/2?sort=itemName`

**Review by mark-prins — COMMENTED** [review]
Just a comment for now - just about to look at the breadcrumb and url.

There's some funk with the list and going back in the browser. If I view a patient record then go back, the list order changes and I can't see the patient which I've just clicked on. That's unsettling!

File: client/packages/common/src/ui/forms/JsonForms/useJsonForms.tsx (hunk 10 above)
Comment (clemens-msupply) [reply client/packages/common/src/ui/forms/JsonForms/useJsonForms.tsx:64]: seems fine here...

File: client/packages/common/src/ui/forms/JsonForms/useJsonForms.tsx (hunk 11 above)
Comment (clemens-msupply) [reply client/packages/common/src/ui/forms/JsonForms/useJsonForms.tsx:114]: I started a branch from this PR and need to make some changes regarding documentRegistry. I fixed it along the way as well, so better wait for my PR...

**Review by mark-prins — COMMENTED** [review]
Have committed a breadcrumb / URL suggestion with 90d55dd38
There were a few files to add, so it was hard to put in as a comment

File: client/packages/system/src/Patient/ListView/ListView.tsx (hunk 13 above)
Comment (CarlosNZ) [reply client/packages/system/src/Patient/ListView/ListView.tsx]: Yes, I agree that is nicer. The only reason I haven't done it like that so far is that it would require two database queries, which I've been trying to avoid initially.

We'd need to look up the document name from the patient record, then pass that name into JsonForms to fetch the document. Yes, we can get document data directly by drilling down into Patient node, but the JsonForms module is supposed to be agnostic to what kind of document it's looking for, so shouldn't know to query the Patient node.

Here's a couple of possible options I can think of to deal with it, you may have some other suggestions:

1. We don't do database lookups in `useJsonForms`. The container element would be responsible for fetching the document and schemas, then just pass them into `useJsonForms` for display. It makes saving more inconsistent though, as the "save" method should probably be in `useJsonForms` as it's super-generic and applies to documents of any type.
2. Just do two queries as mentioned above.
3.  Do a single database lookup in `useJsonForms`, but the container component passes in the database query method (and have a default query where it gets it from "document" node if none provided, which would be the current query)

I can see pros and cons to each, though option 2 is probably the most straightforward. Maybe chat about it on Monday?

File: client/packages/common/src/ui/forms/JsonForms/api/hooks/document/useDocument.ts [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+
+  return useQuery(
+    api.keys.detail(docName),
+    () => api.get.byDocName(docName),
+    // Don't refetch when the edit modal opens, for example. But, don't cache
+    // data when this query is inactive. For example, when navigating away from
+    // the page and back again, refetch.
+    {
+      refetchOnMount: false,
+      cacheTime: 0,
```
Comment (josh-griffin) [inline client/packages/common/src/ui/forms/JsonForms/api/hooks/document/useDocument.ts:15]: think using `staleTime: Infinity` will be helpful. Should prevent background refetches (might want to double check refetching on window focus which is disabled in dev in `host`.. gotcha!) which will result in overriding the current draft when the state (`data`) is synced in `useJSONForms` 😁

**Comment by CarlosNZ** [comment]
> There's some funk with the list and going back in the browser. If I view a patient record then go back, the list order changes and I can't see the patient which I've just clicked on. That's unsettling!

Hmm, this seems to be a problem with the list sorting in general. If you take any table view, sort by a different column, then paste that url into a new tab, it'll revert back to the default sort column. We seem to have broken this somewhere else :/

**Comment by CarlosNZ** [comment]
Merging so @clemens-msupply can continue with "New patient" UI. Some remaining issues can be re-addressed once we have a bit more in place.
